src/Avalonia.Visuals/Rendering/SceneGraph/DeferredDrawingContextImpl.cs

```diff
@@ -0,0 +1,391 @@
+﻿// Copyright (c) The Avalonia Project. All rights reserved.
+// Licensed under the MIT license. See licence.md file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using Avalonia.Media;
+using Avalonia.Platform;
+using Avalonia.VisualTree;
+
+namespace Avalonia.Rendering.SceneGraph
+{
+    /// <summary>
+    /// A drawing context which builds a scene graph.
+    /// </summary>
+    internal class DeferredDrawingContextImpl : IDrawingContextImpl
+    {
+        private readonly ISceneBuilder _sceneBuilder;
+        private VisualNode _node;
+        private int _childIndex;
+        private int _drawOperationindex;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="DeferredDrawingContextImpl"/> class.
+        /// </summary>
+        /// <param name="sceneBuilder">
+        /// A scene builder used for constructing child scenes for visual brushes.
+        /// </param>
+        /// <param name="layers">The scene layers.</param>
+        public DeferredDrawingContextImpl(ISceneBuilder sceneBuilder, SceneLayers layers)
+        {
+            _sceneBuilder = sceneBuilder;
+            Layers = layers;
+        }
+
+        /// <inheritdoc/>
+        public Matrix Transform { get; set; } = Matrix.Identity;
+
+        /// <summary>
+        /// Gets the layers in the scene being built.
+        /// </summary>
+        public SceneLayers Layers { get; }
+
+        /// <summary>
+        /// Informs the drawing context of the visual node that is about to be rendered.
+        /// </summary>
+        /// <param name="node">The visual node.</param>
+        /// <returns>
+        /// An object which when disposed will commit the changes to visual node.
+        /// </returns>
+        public UpdateState BeginUpdate(VisualNode node)
+        {
+            Contract.Requires<ArgumentNullException>(node != null);
+
+            if (_node != null)
+            {
+                if (_childIndex < _node.Children.Count)
+                {
+                    _node.ReplaceChild(_childIndex, node);
+                }
+                else
+                {
+                    _node.AddChild(node);
+                }
+
+                ++_childIndex;
+            }
+
+            var state = new UpdateState(this, _node, _childIndex, _drawOperationindex);
+            _node = node;
+            _childIndex = _drawOperationindex = 0;
+            return state;
+        }
+
+        /// <inheritdoc/>
+        public void Clear(Color color)
+        {
+            // Cannot clear a deferred scene.
```

Comment:
Yeah we could add a separate clear operation, but it's unclear how it would work with layers... Something we should discuss apart from the PR if it's necessary IMO.

But yeah, I think a more sensible move for now is to remove `Clear` from `IDrawingContext`. WPF for example doesn't have it there.